### ISSUE TITLE
Feature Request: Upgrade to Torch 2.9.1+cu129

### ISSUE THREAD

**Issue by piyifan123** [ByteDance-Seed/VeOmni#439]
### Feature request

- [x] Add optional dependency for `torch==2.9.1+cu129`
- [x] Pin/confirm FA2 compatible build for torch 2.9 + CUDA 12.9
- [x] Pin/confirm FA3 compatible build for torch 2.9 + CUDA 12.9
- [x] Fix torch 2.9 regressions (incl. SP collator deprecation)
- [x] Update CI image to CUDA 12.9
- [x] Run CI with torch 2.9.1+cu129
- [x] Promote torch 2.9.1+cu129 to default
- [ ] Clean up transitional code/docs

Note:
- There is no official torch2.9 compatible prebuilt wheels for FA2 and FA3, so we will use the following 3P prebuilt wheels:
   -  FA2: https://github.com/mjun0812/flash-attention-prebuild-wheels
   -  FA3: https://github.com/windreamer/flash-attention3-wheels

#### SP collator using a non-tuple sequence for multidimensional indexing

```
/opt/tiger/modelchef/submodules/Open-VeOmni/veomni/distributed/sequence_parallel/data.py:86: UserWarning: Using a non-tuple sequence for multidimensional indexing is deprecated and will be changed in pytorch 2.9; use x[tuple(seq)] instead of x[seq]. In pytorch 2.9 this will be interpreted as tensor index, x[torch.tensor(seq)], which will result either in an error or a different result (Triggered internally at /pytorch/torch/csrc/autograd/python_variable_indexing.cpp:306.)
  return x[slc].contiguous()
```

### Motivation

We should upgrade VeOmni to **PyTorch 2.9.1 + CUDA 12.9 (cu129 wheels)** for two reasons:

1. **Policy / freshness**: Align with our **-1 policy** now that **Torch 2.10** has been published recently. Torch 2.9.1 is the latest “-1” target and should be supported as a first-class configuration.
2. **Kernel enablement**: Move from **CUDA 12.8 → 12.9** so we can reliably use **cuTE-based kernels** (e.g. **FlashAttention-4 (FA4)** and **SonicMoE**).

### Your contribution

TBA

**Comment by piyifan123** [comment]
Done.
